binderhub/repoproviders.py

```diff
@@ -356,3 +364,68 @@ def get_resolved_ref(self):
 
     def get_build_slug(self):
         return '{user}-{repo}'.format(user=self.user, repo=self.repo)
+
+
+class GistRepoProvider(GitHubRepoProvider):
+    """GitHub gist provider.
+
+    Users must provide a spec that matches the following form (similar to github)
+
+    <username>/<gist-id>[/<ref>]
+
+    The ref is optional, valid values are
+        - a full sha1 of a ref in the history
+        - master
+    If master or no ref is specified the latest revision will be used.
+    """
+
+    allow_secret_gist = Bool(
+        default_value=False,
+        config=True,
+    )
+
+    def __init__(self, *args, **kwargs):
+        # We dont need to initialize entirely the same as github
+        super(RepoProvider, self).__init__(*args, **kwargs)
+        parts = self.spec.split('/')
+        self.user, self.gist_id, *_ = parts
+        if len(parts) > 2:
+            self.unresolved_ref = parts[2]
+        else:
+            self.unresolved_ref = ''
+
+    def get_repo_url(self):
+        return f'https://gist.github.com/{self.gist_id}.git'
+
+    @gen.coroutine
+    def get_resolved_ref(self):
+        if hasattr(self, 'resolved_ref'):
+            return self.resolved_ref
+
+        api_url = f"https://api.github.com/gists/{self.gist_id}"
+        self.log.debug("Fetching %s", api_url)
+
+        resp = yield self.github_api_request(api_url)
+        if resp is None:
+            return None
+
+        ref_info = json.loads(resp.body.decode('utf-8'))
+
+        if (not self.allow_secret_gist) and (not ref_info['public']):
+            raise ValueError("gist is marked as secret, this is unsupported")
```

Comment:
-> "Gist is marked as secret, to enable secret Gist support set GistRepoProvider.allow_secret_gist = True"?

Reply:
Let's add a comment above line 382 or in the docstring that calls out the default behavior too.

For ValueError message:

You seem to want to use a secret Gist, but do not have permission to do so. To enable secret Gist support, set (or have an administrator set) 'GistRepoProvider.allow_secret_gist = True'